### ISSUE TITLE
D3D12: Worms Revolution (Shader errors)

### ISSUE THREAD

**Issue by raven02** [RPCS3/rpcs3#2049]
It crashes out with following shader error in DX12 baclend.

```
rsx::thread} RSX: FS build failed:C:\Users\Ventura\Desktop\Juegos\Emuladores\RPCS3 (PS3)\FragmentProgram.hlsl(136,73): error X3000: syntax error: unexpected token '.'
C:\Users\Ventura\Desktop\Juegos\Emuladores\RPCS3 (PS3)\FragmentProgram.hlsl(136,69-78): error X3013: 'abs': no matching 0 parameter intrinsic function
C:\Users\Ventura\Desktop\Juegos\Emuladores\RPCS3 (PS3)\FragmentProgram.hlsl(136,69-78): error X3013: Possible intrinsic functions are:
C:\Users\Ventura\Desktop\Juegos\Emuladores\RPCS3 (PS3)\FragmentProgram.hlsl(136,69-78): error X3013:     abs(float|half|double|min10float|min16float|int|uint|min12int|min16int|min16uint)
```

**Comment by MarioSonic2987** [comment]
In https://github.com/RPCS3/rpcs3/commit/b4e3d3dbfdbfacdbbeadf447ebe8f8ebc8b45f32, it stucks on this stage in OpenGL. The same happens in DX12 and Vulkan.

![](http://i.imgur.com/j4RMNUB.png)

**Comment by MarioSonic2987** [comment]
@kd-11 @raven02 
![](http://i.imgur.com/rrDxIAD.png)
All stuck in this screen.

Vulkan:
```
·E {rsx::thread} RSX: Src type 3 used, please report this to a developer.
·E {rsx::thread} RSX: Unknown/illegal instruction: 0x26 (forced unit 0)
·E {PPU[0x70000002] Thread (XPlatform::SwitchPs3::Main) [0x00f9d874]} sys_memory: sys_memory_get_page_attribute(addr=0xd002b2e0, attr=*0xd002b240)
·E {rsx::thread} RSX: Warning, version 420 is not yet complete; most version-specific features are present, but some are missing.
ERROR: 0:76: '' :  syntax error
ERROR: 1 compilation errors.  No code generated.


·E {rsx::thread} RSX: 
·F {rsx::thread} class std::runtime_error thrown: Failed to compile fragment shader
(in file C:\rpcs3\rpcs3\Emu\RSX\VK\VKFragmentProgram.cpp:475)
```
DX12:
```
·E {rsx::thread} RSX: Src type 3 used, please report this to a developer.
·E {rsx::thread} RSX: FS build failed:0000000005028d40
·! {rsx::thread} RSX: Add program :
·! {rsx::thread} RSX: *** vp id = 2
·! {rsx::thread} RSX: *** fp id = 24
·F {rsx::thread} class std::runtime_error thrown: fragment program compilation failure
(in file c:\rpcs3\rpcs3\emu\rsx\d3d12\D3D12PipelineState.h:162)
```
OGL:
```
·F {rsx::thread} class gl::glsl::link_exception thrown: linkage failed: 'Fragment info
-------------
0(296) : error C0000: syntax error, unexpected reserved word "discard", expecting ')' at token "discard"
0(297) : error C0000: syntax error, unexpected reserved word "discard", expecting ')' at token "discard"
0(298) : error C0000: syntax error, unexpected reserved word "discard", expecting ')' at token "discard"
0(299) : error C0000: syntax error, unexpected reserved word "discard", expecting ')' at token "discard"
(0) : error C2003: incompatible options for link
'
```
[Shaderlog DX12](https://drive.google.com/open?id=0B9LymxJkpT48RFlBaEVwV3dEUXM)
[Shaderlog OGL](https://drive.google.com/open?id=0B9LymxJkpT48S2UxZ3YzbXdPVHM)
[Shaderlog Vulkan](https://drive.google.com/open?id=0B9LymxJkpT48MlpqNjZlSGhtTTA)

**Comment by MarioSonic2987** [comment]
This issue should be renamed to **DX12/GL/Vulkan: Worms Revolution**.

**Comment by MarioSonic2987** [comment]
Tested in https://github.com/RPCS3/rpcs3/commit/c0589cab05738586089ca70c4ff1d48458f5951d.

Reaches further, but input isn't detected.

**DX12** (Stucks in loading screen with):
```
·F 0:01:49.809368 {rsx::thread} class std::runtime_error thrown: HRESULT = La instancia de dispositivo de GPU se ha suspendido. Use GetDeviceRemovedReason para averiguar cu⬠es la acci󮠡decuada.
(in file C:\rpcs3\rpcs3\Emu\RSX\D3D12\D3D12RenderTargetSets.cpp:279)
```
![](http://i.imgur.com/IWKd133.png)
**OGL**:
![](http://i.imgur.com/8Txs70p.jpg)
**Vulkan**:
![](http://i.imgur.com/lfgyh4K.jpg)

**Comment by kd-11** [comment]
Its possible broken shader code was being loaded before and this was fixed when image_in was updated. Vulkan shading bug and DX12 crash are still valid issues though.

**Comment by MarioSonic2987** [comment]
@kd-11 
[Shaderlog GL](https://drive.google.com/open?id=0B9LymxJkpT48a1g2dEJUdEEzbU0)
[Shaderlog Vulkan](https://drive.google.com/open?id=0B9LymxJkpT48X1ZhMXlUVFBnNkk)
[RenderDoc Vulkan (Title screen)](https://drive.google.com/open?id=0B9LymxJkpT48SmozR3hkQWxNczQ)

**Comment by MarioSonic2987** [comment]
This issue can have opengl label removed, because it doesn't have shader error, despite dimmed lighting.

**Comment by MarioSonic2987** [comment]
Input problem is fixed with PR #2491.
In https://github.com/RPCS3/rpcs3/commit/b70a1edbc7b6067ad3ce038eb7198fa58767576a, shading in Vulkan is solved.
![](http://i.imgur.com/rQvLxyW.jpg)
This can have vulkan label removed.

**Comment by MarioSonic2987** [comment]
With #2631, save problem is fixed. It can mostly go past title screen.
